nn/transformer.py

```diff
@@ -0,0 +1,301 @@
+    self.norm_first = norm_first
+    self.norm_eps = norm_eps
+    self.activation = activation
+    self.dropout = dropout
+    self.att_dropout = att_dropout
+
+  def forward(self, x: nn.LayerRef, *, memory: nn.LayerRef, state: nn.LayerState) -> Tuple[nn.LayerRef, nn.LayerState]:
+    """
+    Two possible forward variants of decoder, defined by self.norm_first, x and memory have shape {B, T, F}
+    """
+    if self.norm_first:
+      x, new_state = x + self._self_attention_block(self.norm(x, epsilon=self.norm_eps), state=state)
+      x = x + self._multi_head_attention_block(self.norm(x, epsilon=self.norm_eps), memory)
+      x = x + self._feed_forward_block(self.norm(x, epsilon=self.norm_eps))
+    else:
+      x, new_state = self.norm(x + self._self_attention_block(x, state=state), epsilon=self.norm_eps)
+      x = self.norm(x + self._multi_head_attention_block(x, memory), epsilon=self.norm_eps)
+      x = self.norm(x + self._feed_forward_block(x), epsilon=self.norm_eps)
+
+    return x, new_state
+
+  def initial_state(self) -> nn.LayerState:
+    """
+    initial state declaration
+    """
+    return nn.LayerState(
+      self_attention=self.self_attn.initial_state(),
+      attention=self.multihead_attn.initial_state())
+
+  def _self_attention_block(self, x: nn.LayerRef, *, state: nn.LayerState) -> Tuple[nn.LayerRef, nn.LayerState]:
+    x, new_state = self.self_attn(x, state=state)
+    return nn.dropout(x, self.dropout), new_state
+
+  def _multi_head_attention_block(self, x: nn.LayerRef, mem: nn.LayerRef) -> nn.LayerRef:
+    x = self.multihead_attn(x, mem, mem)
+    return nn.dropout(x, self.dropout)
+
+  def _feed_forward_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.linear_ff(x)
+    x = self.activation(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    x = self.linear_out(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    return x
+
+
+class TransformerDecoderStep(nn.Module):
+  """
+  Defines the full Decoder of the standard transformer
+  """
+  def __init__(self, decoder_layer: Union[TransformerDecoderLayerStep, Any], num_layers: int,
+               norm=nn.layer_norm, norm_eps: float = 1e-5):
+      """
+      :param decoder_layer: Decoder layer to be stacked num_layers times
+      :param num_layers: Number of layers
+      :param norm: normalization function for output layer normalization
+      :param norm_eps: Epsilon value for output layer normalization
+      """
+      super().__init__()
+      import copy
+      self.layers = nn.ModuleList([copy.deepcopy(decoder_layer) for _ in range(num_layers)])
+
+      self.num_layers = num_layers
+      self.norm = norm
+      self.norm_eps = norm_eps
+
+  def forward(self, x: nn.LayerRef, *, memory: nn.LayerRef, state: nn.LayerState) -> Tuple[nn.LayerRef, nn.LayerState]:
+    """
+    Applies every decoder layer initialized in self.layers.
+    """
+    output = x
+
+    for mod in self.layers:
+      output, state = mod(output, memory=memory, state=state)
```

Comment:
This is wrong. You don't pass the same state through all the modules.
Every module has an own state.
```suggestion
    for key, mod in self.layers.named_children():
      output, state[key] = mod(output, memory=memory, state=state[key])
```